### PR TITLE
Issue #5005: increased pitest coverage for metrics package to 100%

### PR DESCRIPTION
**Pull request by rnveach**
Issue #5005

100% coverage.

### CODE REVIEW

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/npathcomplexity/InputNPathComplexityDefault.java [hunk 1]
```diff
@@ -109,4 +109,17 @@ public void run() {
         };
         new Thread(runnable).start();       
     }
+
+    public void InputNestedTernaryCheck() {
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/npathcomplexity/InputNPathComplexityDefault.java]: This case proves why `lineNo == endLineNo` can't be removed.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheckTest.java [hunk 2]
```diff
@@ -248,6 +251,25 @@ public void testDefaultHooks() {
         Assert.assertEquals("No exception messages expected", 0, messages2.size());
     }
 
+    /**
+     * This must be a reflection test as it is too difficult to hit normally and
+     * the responsible code can't be removed without failing tests.
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheckTest.java]: Comment explains why we must use reflection.

**Comment by romani** [comment]
@rnveach ,

did you try to change a code as pitest and run regression ?

screenshot_2017-12-08_16-52-53
```
1. changed conditional boundary → KILLED 
2. changed conditional boundary → SURVIVED 
3. negated conditional → KILLED 
4. negated conditional → KILLED 
5. negated conditional → KILLED
```

it is not completely clear where is survived , but we have only two options. So two diff reports.
Is it too much to prove that reflection is required ?

**Comment by rnveach** [comment]
@romani 
https://github.com/checkstyle/checkstyle/issues/5005#issuecomment-350332862
`Report is hard to understand, but the conditional boundary surviving is columnNo > endColumnNo`

`Conditional boundary` is changing `>` to `>=`, so we must either create/hide a violation during the mutation and it must be at the same line/column position as what we are checking.

https://github.com/checkstyle/checkstyle/issues/5005#issuecomment-350391816
`This can't be removed because this test and line fails`

So our only option is kill the mutation.
`TokenEnd` is only used for `processingTokenEnd` and it is only set during `visitConditional` and `visitUnitaryOperator`. For it to be the same line/column, it must be the exact same token or a token who has the same line/column as it's child and we visit. We never visit the same token twice and we are only visiting on specific tokens. Regression also didn't show me anything we can use.
I am out of options for this one.

**Comment by romani** [comment]
In comments I saw that you looks like tried to remove lines, that definitely caught by UTs.
Did you try to change ">" to ">=" and run regression ?
If no diff - ok to use reflection but please put your last comment into javadoc to be more specific what we tried and why it is so hard.

**Comment by rnveach** [comment]
http://rveach.no-ip.org/checkstyle/regression/reports/139/
https://github.com/rnveach/checkstyle/commits/issue_5005_proof
No differences.

**Comment by rnveach** [comment]
> please put your last comment into javadoc to be more specific 

Done.

**Pull request merged**